app/dcr_monitor/tee_backend/confidential_space.go

```diff
@@ -67,3 +86,238 @@ func (c *TEEProviderGCPConfidentialSpace) CleanUpInstance(instanceName string) e
+			},
+			},
+		},
+		Tags: &computepb.Tags{
+			Items: []string{"tee-instance"},
+		},
+		ServiceAccounts: []*computepb.ServiceAccount{
+			&computepb.ServiceAccount{
+				Email:  &c.saEmail,
+				Scopes: []string{"https://www.googleapis.com/auth/cloud-platform"},
+			},
+		},
+		Name:        &instanceName,
+		MachineType: &machineType,
+		Disks: []*computepb.AttachedDisk{
+			&computepb.AttachedDisk{
+				DiskSizeGb: &diskSize,
+				AutoDelete: proto.Bool(true),
+				Boot:       proto.Bool(true),
+				InitializeParams: &computepb.AttachedDiskInitializeParams{
+					SourceImage: &imageSource,
+				},
+			},
+		},
+		Scheduling: &computepb.Scheduling{
+			OnHostMaintenance: proto.String("TERMINATE"),
+		},
+		NetworkInterfaces: []*computepb.NetworkInterface{
+			&computepb.NetworkInterface{
+				AccessConfigs: []*computepb.AccessConfig{
+					&computepb.AccessConfig{
+						Name: proto.String("external-nat"),
+						Type: proto.String("ONE_TO_ONE_NAT"),
+					},
+				},
+				Network:    &network,
+				Subnetwork: &subNetwork,
+			},
+		},
+		CanIpForward: proto.Bool(false),
+	}
+	req := &computepb.InsertInstanceRequest{
+		InstanceResource: &instanceResource,
+		Zone:             c.zone,
+		Project:          c.projectId,
+	}
+	return req
+}
+
+func (c *TEEProviderGCPConfidentialSpace) updateWorkloadIdentityPoolProvider(name string, imageDigest string) error {
+	requestBody, err := c.workloadIdentityRequestBody(name, imageDigest)
+	if err != nil {
+		return err
+	}
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: false,
+			MinVersion:         tls.VersionTLS12,
+		},
+	}
+	client := &http.Client{Transport: tr}
+	updateWipProviderUrl := fmt.Sprintf(
+		"https://iam.googleapis.com/v1/projects/%s/locations/global/workloadIdentityPools/dcr-%s-pool/providers/%s?updateMask=attributeCondition",
+		c.projectId, c.env, name)
+	req, err := http.NewRequest("PATCH", updateWipProviderUrl, bytes.NewReader(requestBody))
+	if err != nil {
+		return errors.Wrap(err, "failed to create request")
+	}
+	token, err := c.getAccessToken()
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed to do http request")
+	}
+	defer resp.Body.Close()
+	res, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Wrap(err, "failed to read http response")
+	}
+	hlog.Debugf("%v", string(res))
+	return nil
+}
+
+// FIXME: duplicated from pkg/cloud
+type WorkloadIdentityPoolProvider struct {
+	DisplayName        string            `json:"displayName"`
+	Description        string            `json:"description"`
+	AttributeMapping   map[string]string `json:"attributeMapping"`
+	AttributeCondition string            `json:"attributeCondition"`
+	OIDC               OIDC              `json:"oidc"`
+}
+
+// FIXME: duplicated from pkg/cloud
```

Comment:
Do it mean we should remove the duplicated struct in the pkg/cloud

Reply:
I have completely deduplicated it by removing `PrepareResourcesForUser`.